man/man1/9p.1
.TH 9P 1
.SH NAME
9p \- read and write files on a 9P server
.SH SYNOPSIS
.B 9p
[
.I options
]
.B read
[
.B -f
]
.I path
.br
.B 9p
[
.I options
]
.B readfd
.I path
.PP
.B 9p
[
.I options
]
.B write
[
.B -l
]
.I path
.br
.B 9p
[
.I options
]
.B writefd
.I path
.PP
.B 9p
[
.I options
]
.B stat
.I path
.PP
.B 9p
[
.I options
]
.B rdwr
.I path
.PP
.B 9p
[
.I options
]
.B ls
[
.B -dl
]
.I path...
.SH DESCRIPTION
.I 9p
is a trivial 9P client that can access a single file on a 9P server.
It can be useful for manual interaction with a 9P server or for
accessing simple 9P services from within shell scripts.
The common options are:
.TP
.B -a\fI addr
dial
the server at
.I addr
.TP
.B -A\fI aname
attach to the file system named
.I aname
.TP
.B -n
mount without authentication
.PP
The first argument is a command, one of:
.TP
.B read
print the contents of
.I path
to standard output ;
the
.B -f
option causes
.I read
to follow the file, exactly like the
.B -f
option of
.IR tail(1)
.TP
.B write
write data on standard input to
.IR path ;
the
.B -l
option causes
.I write
to write one line at a time
.TP
.BR readfd ", " writefd
like
.B read
and
.B write
but use
.IR openfd (9p)
instead of
.IR open ;
this masks errors and is mainly useful for debugging
the implementation of
.I openfd
.TP
.B stat
execute
.I stat (9p)
on
.I path
and print the result
.TP
.B rdwr
Open
.I path
for reading and writing.
Then repeat until end-of-file on standard input:
copy a line from the file to standard output,
copy a line from standard input to the file.
Print errors, but don't give up.
.B Rdwr
is useful for interacting with servers like
.MR factotum 4 .
.TP
.B ls
Print a directory listing in the format of
.MR ls 1 .
The
.B -d
and
.B -l
flags have the same meaning as in
.IR ls .
.PD
.PP
.I 9p
dials
.I address
to connect to the 9P server.
If the
.B -a
option is not given,
.I 9p
requires the
.I path
to be of the form
.IB service / subpath \fR;
it connects to the Unix domain socket
.I service
in the name space directory
(see
.MR intro 4 )
and then accesses
.IR subpath .
.SH EXAMPLE
To update
.MR plumber 4 's
copy of your plumbing rules after editing
.BR $HOME/lib/plumbing :
.IP
.EX
cat $HOME/lib/plumbing | 9p write plumb/rules
.EE
.PP
To display the contents of the current
.MR acme 4
window:
.IP
.EX
9p read acme/$winid/body
.EE
.SH SOURCE
.B \*9/src/cmd/9p.c
.SH SEE ALSO
.MR intro 4 ,
.IR intro (9p),
.MR 9pclient 3


CHANGES
# Note: the name listed next to a change sometimes
# submitted the patch but more often just reported
# the problem.

# For entries more recent than January 30, 2008,
# see the Git repository history at 
# https://github.com/9fans/plan9port/.

January 30, 2008
	acme: fix regexp match choice bug
	devdraw: add F11 for full-screen toggle
	rio: accomodate full-screen apps more gracefully
	sam: fix regexp match choice bug

January 28, 2008
	xd: fix printing of 4-byte values on 64-bit machines (Sam Hopkins)

January 12, 2008
	delatex: reset lexer state at each beginning of file
	ipso: do not flush factotum keys unless using factotum file
	libdiskfs: fix ext2 directory reading bug (Mike Mammarella)
	lp: move devices to devices.sample
	rio: make full-screen work properly; add showevent
	venti: fewer calls to gettimeofday
	vmount: need to say -t nfs now on Linux

January 10, 2008
	acme: make STACK bigger for OS X (David Swasey)
	dist,unix: use MKSHELL=$PLAN9/bin/rc (John DeGood)
	libdraw: fix spinning event loop (David Bulkow)
	libregexp: fix match choice bug

December 7, 2007
	acme, sam, libregexp: regexp fixes and reversions

December 4, 2007
	hoc: print exact floating point output
	sam: turn off DEBUG logging (oops)
	vbackup: fix uninitialized variable use in -i code

November 28, 2007
	acme: refresh window properly after Tab command (Fazlul Shahriar)

November 27, 2007
	.hgignore: ignore more
	libdiskfs: make ext2 use little-endian disk format on big-endian machines (Joshua Wood)
	acme: accept clicks between body and scroll bar (Axel Belinfante)
	tpic: clean up better
	upas/smtp: remove intermediate files
	vbackup: add -i flag to avoid score prefetching (David Swasey)
	venti: add debugging HTML form

November 25, 2007
	acme: avoid infinite loop on search in 8125-byte file (Roger Peppe)
	lib9: avoid DIOCGMEDIASIZE on FreeBSD 4.11 (Rodrigo Readi)
	mailfs: use root in imapcopylist (Richard Bilson)
	moveplan9.files: add tmac/tmac.an (Tony Lainson)

November 23, 2007
	9pfuse: ignore Linux O_CLOEXEC flag (Michael Kaminsky)

November 22, 2007
	9pfuse: fix array bounds error (Michael Teichgräber)
	vbackup: fix fsysopen error message (Joshua Wood)

November 7, 2007
	9fs: use -n to mount sources (Jeff Sickel),
	9pfuse: MacFUSE moved again! (Jeff Sickel)

November 6, 2007
	.cvsignore: add 9pm (for old trees)
	unvf: ignore case in Content-Transfer-Encoding: line.
	validateattachment: reject malformed zip files

November 5, 2007
	building: use ar s flag to avoid ranlib (OS X)
	diff: continue after i/o errors when diffing many files
	factotum: silence gcc warning
	getns: turn / into _ in $DISPLAY
	libdiskfs: OS X linker fixes
	libmach: OS X linker fixes
	libmp: no text jump labels (OS X)
	libthread: use our own swapcontext on OS X
	mailfs: handle NIL body, add -r flag (Richard Bilson)

November 4, 2007
	man: document hg interface, fix CVS host name

November 3, 2007
	libthread: more OS X 10.5 changes (Jeff Sickel)

October 31, 2007
	venti: use 9ar, not ar (fixes Darwin build)

October 30, 2007
	libthread: OS X 10.5 tweaks (Jeff Sickel, Bakul Shah)

October 29, 2007
	grap: fclose, error context bug fixes (Doug Gwyn)
	venti: fix sync deadlock, add /proc stub

October 12, 2007
	9p: add -n to disable authentication
	9pserve: add -n to reject authentication
	srv: add -n to reject authentication

October 9, 2007
	mount: ask for Unix test explicitly (Chris King)

September 24-25, 2007
	venti: many fixes, new index cache

September 16, 2007
	factotum: fix build error (Axel Belinfante)

September 15, 2007
	auxstats: use acpi for battery on Linux (Andrey Mirtchovski)

September 13, 2007
	factotum: implement 9pcr (of dubious utility)

August 28, 2007
	9pserve: fix infinite loop bug

August 23, 2007
	venti: add -s flag, bug fixes (Michael Kaminsky)

August 22, 2007
	9.rc: change to work with rc -e (Tony Lainson)
	9pserve: fix races causing assert(c->nmsg == 0) to fail.
	acme: fix warning memory leak (Michael Teichgräber)
	libfmt: add fmtlocale to Makefile (Faried Nawaz)
	libthread: simplify alt tracking, possibly fixing bug
	mailfs: fix handling of FLAGS responses (Robert Vollmert)
	unix ports: Linux x86-64 Makefiles (Ævar Arnfjörð Bjarmason)

July 25, 2007
	9pfuse: even more MacFUSE path changes

June 29, 2007
	dict: look in $dictpath for dictionaries

June 24, 2007
	mailfs: fix subject line in header file.

June 23, 2007
	mailfs: various IMAP bug fixes
	Mail: add Get command, refresh at startup,
		fix deleted messages bug
	nedmail: refresh at startup, on i and y.

June 20, 2007
	9pfuse: add -a aname to get at sourcesdump

June 19, 2007
	9pfuse: the MacFUSE file locations changed again (Rob Pike)

June 18, 2007
	upas/marshal: remove debugging print
	9pserve: ignore window resizes (fix really should be elsewhere)
	libdiskfs: more HFS fixes (David Swasey)

June 15, 2007
	acme(1): fix Snarf bug, change Indent ON to apply to all windows.

June 12, 2007
	vcat: always write last block, to make disk correct size
	vbackup: more OS X fixes, add -M flag
	vbackup(8): document -M

June 9, 2007
	libregexp, sam, acme: day 1 bug in sam regexp code

June 8, 2007
	libdiskfs: add HFS resource fork support (David Swasey)

June 5, 2007
	ramfs: be less chatty

June 3, 2007
	vbackup: fix dotdot (David Swasey)

May 29, 2007
	libdiskfs: more HFS fixes (David Swasey)
	
May 28, 2007
	libdiskfs: missing Bterm, make ext2 quieter,
		add disk partitions, HFS code (David Swasey)

May 21, 2007
	libregexp: fix pointer comparison warning (Ramon de Vera)
	test: fix
	
May 20, 2007
	touch: print error if wstat fails
	libc: use futimes on Linux to implement dirfwstat

May 19, 2007
	test: import from Plan 9

May 18, 2007
	libregexp: fix bug about leading ^ in regexp on multiline text (Roger Peppe)

May 15, 2007
	libc: new media size code for _p9dir on FreeBSD (Lou Kamenov)
	acme(4): document tab width field in ctl file

May 10, 2007
	libc: more NaN fixes (rsc)
		use device names in _p9dir for FreeBSD (Lou Kamenov)

src/cmd/9p.c
#include <u.h>
#include <signal.h>
#include <libc.h>
#include <bio.h>
#include <fcall.h>
#include <9pclient.h>
#include <auth.h>
#include <thread.h>
#include <poll.h>

char *addr;

void
usage(void)
{
	fprint(2, "usage: 9p [-n] [-a address] [-A aname] cmd args...\n");
	fprint(2, "possible cmds:\n");
	fprint(2, "	read name\n");
	fprint(2, "	readfd name\n");
	fprint(2, "	write [-l] name\n");
	fprint(2, "	writefd name\n");
	fprint(2, "	stat name\n");
	fprint(2, "	rdwr name\n");
	fprint(2, "	ls [-ldn] name\n");
	fprint(2, "without -a, name elem/path means /path on server unix!$ns/elem\n");
	threadexitsall("usage");
}

CFsys *(*nsmnt)(char*, char*) = nsamount;
CFsys *(*fsmnt)(int, char*) = fsamount;

char *aname;
void xread(int, char**);
void xwrite(int, char**);
void xreadfd(int, char**);
void xwritefd(int, char**);
void xstat(int, char**);
void xls(int, char**);
void xrdwr(int, char**);
void xrm(int, char**);
void xcreate(int, char**);
void xcon(int, char**);

struct {
	char *s;
	void (*f)(int, char**);
} cmds[] = {
	"con", xcon,
	"read", xread,
	"write", xwrite,
	"readfd", xreadfd,
	"writefd", xwritefd,
	"stat", xstat,
	"rdwr", xrdwr,
	"ls", xls,
	"rm", xrm,
	"create", xcreate,
};

void
threadmain(int argc, char **argv)
{
	char *cmd;
	int i;

	ARGBEGIN{
	case 'A':
		aname = EARGF(usage());
		break;
	case 'a':
		addr = EARGF(usage());
		if(strchr(addr, '!') == nil)
			addr = netmkaddr(addr, "tcp", "9fs");
		break;
	case 'n':
		nsmnt = nsmount;
		fsmnt = fsmount;
		break;
	case 'D':
		chatty9pclient = 1;
		break;
	default:
		usage();
	}ARGEND

	signal(SIGINT, SIG_DFL);

	if(argc < 1)
		usage();

	cmd = argv[0];
	for(i=0; i<nelem(cmds); i++){
		if(strcmp(cmds[i].s, cmd) == 0){
			cmds[i].f(argc, argv);
			threadexitsall(0);
		}
	}
	usage();
}

CFsys*
xparse(char *name, char **path)
{
	int fd;
	char *p;
	CFsys *fs;

	if(addr == nil){
		p = strchr(name, '/');
		if(p == nil)
			p = name+strlen(name);
		else
			*p++ = 0;
		*path = p;
		fs = nsmnt(name, aname);
		if(fs == nil)
			sysfatal("mount: %r");
	}else{
		*path = name;
		if((fd = dial(addr, nil, nil, nil)) < 0)
			sysfatal("dial: %r");
		if((fs = fsmnt(fd, aname)) == nil)
			sysfatal("mount: %r");
	}
	return fs;
}

CFid*
xopen(char *name, int mode)
{
	CFid *fid;
	CFsys *fs;

	fs = xparse(name, &name);
	fid = fsopen(fs, name, mode);
	if(fid == nil)
		sysfatal("fsopen %s: %r", name);
	return fid;
}

int
xopenfd(char *name, int mode)
{
	CFsys *fs;

	fs = xparse(name, &name);
	return fsopenfd(fs, name, mode);
}

void
checkout()
{
	int nfd;
	struct pollfd *pfds;

	nfd = 2;
	pfds = calloc(nfd, sizeof(struct pollfd));
	pfds[0].fd = 1;
	pfds[1].fd = 2;
	for(;;) {
		if (poll(pfds, nfd, -1) < 0)
			sysfatal("poll");
		if (pfds[0].revents & POLLERR || pfds[1].revents & POLLERR)
			threadexitsall(0);
	}
}

void
xread(int argc, char **argv)
{
	char buf[4096];
	int n;
	CFid *fid;

	ARGBEGIN{
	default:
		usage();
	}ARGEND

	if(argc != 1)
		usage();

	fid = xopen(argv[0], OREAD);
	proccreate(checkout, nil, 32768);
	while((n = fsread(fid, buf, sizeof buf)) > 0)
		if(write(1, buf, n) < 0)
			sysfatal("write error: %r");
	fsclose(fid);
	if(n < 0)
		sysfatal("read error: %r");
	threadexitsall(0);
}

void
xreadfd(int argc, char **argv)
{
	char buf[4096];
	int n;
	int fd;

	ARGBEGIN{
	default:
		usage();
	}ARGEND

	if(argc != 1)
		usage();

	fd = xopenfd(argv[0], OREAD);
	while((n = read(fd, buf, sizeof buf)) > 0)
		if(write(1, buf, n) < 0)
			sysfatal("write error: %r");
	if(n < 0)
		sysfatal("read error: %r");
	threadexitsall(0);
}

void
xwrite(int argc, char **argv)
{
	char buf[4096];
	int n, did;
	CFid *fid;
	Biobuf *b;
	char *p;
	int byline;

	byline = 0;
	ARGBEGIN{
	case 'l':
		byline = 1;
		break;
	default:
		usage();
	}ARGEND

	if(argc != 1)
		usage();

	did = 0;
	fid = xopen(argv[0], OWRITE|OTRUNC);
	if(byline){
		n = 0;
		b = malloc(sizeof *b);
		if(b == nil)
			sysfatal("out of memory");
		Binit(b, 0, OREAD);
		while((p = Brdstr(b, '\n', 0)) != nil){
			n = strlen(p);
			did = 1;
			if(fswrite(fid, p, n) != n)
				fprint(2, "write: %r\n");
		}
		free(b);
	}else{
		while((n = read(0, buf, sizeof buf)) > 0){
			did = 1;
			if(fswrite(fid, buf, n) != n)
				sysfatal("write error: %r");
		}
	}
	if(n == 0 && !did){
		if(fswrite(fid, buf, 0) != 0)
			sysfatal("write error: %r");
	}
	if(n < 0)
		sysfatal("read error: %r");
	fsclose(fid);
	threadexitsall(0);
}

void
xwritefd(int argc, char **argv)
{
	char buf[4096];
	int n;
	int fd;

	ARGBEGIN{
	default:
		usage();
	}ARGEND

	if(argc != 1)
		usage();

	fd = xopenfd(argv[0], OWRITE|OTRUNC);
	while((n = read(0, buf, sizeof buf)) > 0)
		if(write(fd, buf, n) != n)
			sysfatal("write error: %r");
	if(n < 0)
		sysfatal("read error: %r");
	threadexitsall(0);
}

void
xstat(int argc, char **argv)
{
	Dir *d;
	CFsys *fs;
	char *name;

	ARGBEGIN{
	default:
		usage();
	}ARGEND

	if(argc != 1)
		usage();

	name = argv[0];
	fs = xparse(name, &name);
	if((d = fsdirstat(fs, name)) == 0)
		sysfatal("dirstat: %r");
	fmtinstall('D', dirfmt);
	fmtinstall('M', dirmodefmt);
	print("%D\n", d);
	threadexitsall(0);
}

void
xrdwr(int argc, char **argv)
{
	char buf[4096];
	char *p;
	int n;
	CFid *fid;
	Biobuf *b;

	ARGBEGIN{
	default:
		usage();
	}ARGEND

	if(argc != 1)
		usage();

	if((b = Bfdopen(0, OREAD)) == nil)
		sysfatal("out of memory");
	fid = xopen(argv[0], ORDWR);
	for(;;){
		fsseek(fid, 0, 0);
		if((n = fsread(fid, buf, sizeof buf)) < 0)
			fprint(2, "read: %r\n");
		else{
			if(write(1, buf, n) < 0 || write(1, "\n", 1) < 0)
				sysfatal("write error: %r");
		}
		if((p = Brdstr(b, '\n', 1)) == nil)
			break;
		n = strlen(p);
		if(fswrite(fid, p, n) != n)
			fprint(2, "write: %r\n");
		free(p);
	}
	fsclose(fid);
	Bterm(b);
	threadexitsall(0);
}

void
xcreate(int argc, char **argv)
{
	int i;
	CFsys *fs;
	CFid *fid;
	char *p;

	ARGBEGIN{
	default:
		usage();
	}ARGEND

	if(argc == 0)
		usage();

	for(i=0; i<argc; i++){
		fs = xparse(argv[i], &p);
		if((fid=fscreate(fs, p, OREAD, 0666)) == nil)
			fprint(2, "create %s: %r\n", argv[i]);
		else
			fsclose(fid);
		fsunmount(fs);
	}
}

void
xrm(int argc, char **argv)
{
	int i;
	CFsys *fs;
	char *p;

	ARGBEGIN{
	default:
		usage();
	}ARGEND

	if(argc == 0)
		usage();

	for(i=0; i<argc; i++){
		fs = xparse(argv[i], &p);
		if(fsremove(fs, p) < 0)
			fprint(2, "remove %s: %r\n", argv[i]);
		fsunmount(fs);
	}
}

void
rdcon(void *v)
{
	int n;
	char buf[4096];
	CFid *fid;

	fid = v;
	for(;;){
		n = read(0, buf, sizeof buf);
		if(n <= 0)
			threadexitsall(0);
		if(buf[0] == 'R'-'A'+1)
			threadexitsall(0);
		if(fswrite(fid, buf, n) != n)
			fprint(2, "write: %r\n");
	}
}

void
xcon(int argc, char **argv)
{
	char buf[4096], *r, *w, *e;
	int n, nocr;
	CFid *fid;

	nocr = 1;

	ARGBEGIN{
	case 'r':
		nocr = 0;
		break;
	default:
		usage();
	}ARGEND

	if(argc != 1)
		usage();

	fid = xopen(argv[0], ORDWR);
	proccreate(rdcon, fid, 32768);
	for(;;){
		n = fsread(fid, buf, sizeof buf);
		if(n <= 0)
			threadexitsall(0);
		if(nocr){
			for(r=w=buf, e=buf+n; r<e; r++)
				if(*r != '\r')
					*w++ = *r;
			n = w-buf;
		}
		if(write(1, buf, n) != n)
			threadexitsall(0);
	}
	fsclose(fid);
	threadexitsall(0);
}

static char *mon[] =
{
	"Jan", "Feb", "Mar", "Apr", "May", "Jun",
	"Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
};


int
timefmt(Fmt *fmt)
{
	ulong u;
	static ulong time0;
	Tm *tm;

	if(time0 == 0)
		time0 = time(0);
	u = va_arg(fmt->args, ulong);
	tm = localtime(u);
	if((long)(time0-u) < 6*30*86400)
		return fmtprint(fmt, "%s %2d %02d:%02d",
			mon[tm->mon], tm->mday, tm->hour, tm->min);
	return fmtprint(fmt, "%s %2d %5d",
		mon[tm->mon], tm->mday, tm->year+1900);
}

static int
dircmp(const void *va, const void *vb)
{
	Dir *a, *b;

	a = (Dir*)va;
	b = (Dir*)vb;
	return strcmp(a->name, b->name);
}

static int
timecmp(const void *va, const void *vb)
{
	Dir *a, *b;

	a = (Dir*)va;
	b = (Dir*)vb;
	if(a->mtime < b->mtime)
		return -1;
	else if(a->mtime > b->mtime)
		return 1;
	else
		return 0;
}

char *dot[] = { "." };

void
xls(int argc, char **argv)
{
	char *err, *name, *xname, *f[4], buf[4096];
	int i, j, l, sort;
	int lflag, dflag, tflag, n, len[4];
	Dir *d;
	CFid *fid;
	CFsys *fs;

	err = nil;
	sort = 1;
	lflag = dflag = tflag = 0;
	ARGBEGIN{
	case 'n':
		sort = 0;
		break;
	case 'l':
		lflag = 1;
		break;
	case 'd':
		dflag = 1;
		break;
	case 't':
		tflag = 1;
		break;
	}ARGEND

	fmtinstall('D', dirfmt);
	fmtinstall('M', dirmodefmt);
	quotefmtinstall();
	fmtinstall('T', timefmt);

	if(argc == 0){
		argv = dot;
		argc = 1;
	}
	for(i=0; i<argc; i++){
		name = argv[i];
		fs = xparse(name, &xname);
		if((d = fsdirstat(fs, xname)) == nil){
			fprint(2, "dirstat %s: %r\n", name);
			fsunmount(fs);
			err = "errors";
			continue;
		}
		if((d->mode&DMDIR) && !dflag){
			if((fid = fsopen(fs, xname, OREAD)) == nil){
				fprint(2, "open %s: %r\n", name);
				fsunmount(fs);
				free(d);
				err = "errors";
				continue;
			}
			free(d);
			n = fsdirreadall(fid, &d);
			fsclose(fid);
			if(n < 0){
				fprint(2, "dirreadall %s: %r\n", name);
				fsunmount(fs);
				err = "errors";
				continue;
			}
			if(sort){
				if(tflag)
					qsort(d, n, sizeof d[0], timecmp);
				else
					qsort(d, n, sizeof d[0], dircmp);
			}
			for(j=0; j<4; j++)
				len[j] = 0;
			for(i=0; i<n; i++){
				d[i].type = 'M';
				d[i].dev = 0;
				snprint(buf, sizeof buf, "%d %s %s %lld",
					d[i].dev, d[i].uid, d[i].gid, d[i].length);
				getfields(buf, f, 4, 0, " ");
				for(j=0; j<4; j++){
					l = strlen(f[j]);
					if(l > len[j])
						len[j] = l;
				}
			}
			for(i=0; i<n; i++){
				if(!lflag){
					print("%q\n", d[i].name);
					continue;
				}
				print("%M %C %*d %*s %*s %*lld %T %q\n",
					d[i].mode, d[i].type, len[0], d[i].dev,
					-len[1], d[i].uid, -len[2], d[i].gid,
					len[3], d[i].length, d[i].mtime, d[i].name);
			}
		}else{
			d->type = 'M';
			d->dev = 0;
			if(lflag)
				print("%M %C %d %s %s %lld %T %q\n",
					d->mode, d->type, d->dev,
					d->uid, d->gid, d->length, d->mtime, d->name);
			else
				print("%q\n", d->name);
		}
		free(d);
	}
	threadexitsall(err);
}
